open-craft/opencraft
# [OC-3157,2517] Handle potentially wasteful load balancer reconfigurations internally

This PR introduces some logic internal to the load balancer API that makes it so that the load balancer is not reconfigured multiple times wastefully/redundantly whenever a call to reconfigure/deconfigure the load balancer is made.

In the process it also updates the mass redeployment command, to take into account this update to the load balancer. Previously, we were trying to be quite careful with not reconfiguring the load balancer for each instance in a batch. Now, we can ignore this, and focus on mass redeployment-related logic only, and let the LB itself handle preventing wasteful reconfigurations.

This also introduces a small task that runs every minute -- it makes sure that any load balancer that requires reconfiguration (marked `dirty` for when potentially redundant/wasteful calls are made to the LB model API for reconfiguration), gets reconfigured, but no others.

**Testing Instructions**:

I think the best way to test the mass redeployment section of this PR would be to just run the mass redeployment command on stage. There are currently 4 instances there. What we should do is run the following command:

```bash
make manage "instance_redeploy --tag oc-3157-2517 --num-attempts 2 --batch-size 2" >& ~/oc-3157-2517.log
```

For one, we're obviously looking to see if the mass redeployment command still works correctly given these changes, and we'll find that out inevitably.

More specifically, we want to know the following:

1. Are new appservers that provision successfully marked active?

We can figure this out by just looking at the GUI and seeing whether the UI shows the latest appserver as being active.

2. Are old appservers deactivated?

We can figure this out by just looking at the GUI and seeing whether the UI shows all appservers besides the latest as being inactive.

---

For the LB portion of these changes, we *could* try and see the LB configuration version after the mass redeployment command succeeds, but I'm not sure exactly what numbers to look for and I think it depends a lot on whether 2 appservers both complete provisioning at roughly the same time (within ~10 minutes of each other).

Instead, we can try this:

1. Spawn new appservers for 2 different instances.
1. Take note of the LB deployed configuration version & the normal configuration version.
1. When provisioning completes for both of them, manually activate 1.
1. Confirm quickly that configuration version > deployed configuration version.
1. Manually activate the other appserver.
1. Confirm that configuration version is now 1 greater than before, and deployed configuration version hasn't changed yet.
1. Wait about 10-15 minutes and come back to see that deployed configuration version is 1 less than the normal configuration version.
1. Wait about 10-15 minutes and come back to see that the values are still the same.

I think the above steps may be overkill, I'm not really sure. Let me know if any of these steps can be squashed / made simpler.
